### ISSUE TITLE
Resuming training with custom scheduler loads wrong learning rate

### ISSUE THREAD

**Issue by rob-hen** [Lightning-AI/lightning#18588]
### Bug description

I implemented a learning rate scheduler with a warmup phase using `LambdaLR.` For a learning rate `self.learnig_rate`, it will start with the learning rate `self.warmup_start_factor * self.learning_rate` and keeps increasing until it reaches the learning rate `self.learnig_rate`. Then the learning rate `self.learnig_rate` is kept. This looks all fine. 
However, when I resume the training from a saved checkpoint, the learning rate is wrong. 

When we resume the training, it initially calls `scale_function(0)` (see my code) during an initial validation with `global_step=0`. Now for the training it uses the learning rate `self.warmup_start_factor(self.trainer.global_step) * scale_function(0) * self.learning_rate` instead of `self.warmup_start_factor(self.trainer.global_step) * self.learning_rate`
    

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import torch


def configure_optimizers(self):
    from torch.optim.lr_scheduler import LambdaLR
    import numpy as np
    scale_factor = np.linspace(self.warmup_start_factor,1,num=1000)
    optimizer = torch.optim.AdamW(
                self.model.parameters(), lr=self.learning_rate)
    def scale_function(step):
           if step < 1000:
               scale = scale_factor[step]
           else:
              scale = 1
           return scale 
    scheduler = LambdaLR(optimizer, lr_lambda=scale_function)
    return ([optimizer], [{"scheduler": scheduler, "interval": "step", "frequency": 1}])

def training_step(self, batch, batch_idx):

     # compute one step....
     if self.trainer.is_global_zero:
         lr = self.optimizers().param_groups[0]['lr']
         self.log("lr", lr, prog_bar=True, logger=True,
                         on_step=True, on_epoch=False, rank_zero_only=True)
```

### Error messages and logs

The learning rate when we resume training is multiplied by `scale_function(0)`. Measurement: `self.optimizers().param_groups[0]['lr']`.

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @borda

**Comment by awaelchli** [comment]
Hi @rob-hen 
Sorry to say, I think the code you posted doesn't make any sense. You posted a scheduler that takes a lambda function that you defined as:

```py
def scale_function(step):
           if step < 1000:
               scale = scale_factor[step]
           else:
              scale = 1
           return lr 
```
This function returns a constant, the variable `lr`, and does not depend on the step input. Furthermore, the `lr` variable isn't even bound. This will not work. 

Here is a full, runnable, example using a LambdaLR scheduler. This code example demonstrates how the learning rate correctly decreases, and resumes decreasing after we resume from a checkpoint. 

Since you couldn't post any working code that we can debug, please consider taking my example and modifying it to your use case to show the problem. A reproducible runnable code example would be very much appreciated, thank you for reporting!


```py
import torch
from torch.optim.lr_scheduler import LambdaLR
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()

        print(
            f"lr: {self.trainer.lr_scheduler_configs[0].scheduler.get_last_lr()[0]:.4f}",
            f"lr_opt: {self.trainer.optimizers[0].param_groups[0]['lr']:.4f}"
        )
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        scheduler = LambdaLR(optimizer, lr_lambda=lambda epoch: 0.95 ** epoch)
        return [optimizer], [{"scheduler": scheduler, "interval": "step", "frequency": 1}]


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=8)

    model = BoringModel()
    trainer = Trainer(max_epochs=1, enable_model_summary=False, enable_progress_bar=False)
    trainer.fit(model, train_dataloaders=train_data)

    print("Resume")
    model = BoringModel()
    trainer2 = Trainer(max_epochs=2, enable_model_summary=False, enable_progress_bar=False)
    trainer2.fit(model, train_dataloaders=train_data, ckpt_path=trainer.checkpoint_callback.best_model_path)


if __name__ == "__main__":
    run()

```

**Comment by rob-hen** [comment]
Hi @awaelchli, 

thank you for your detailed answer.
Regarding my post, there was indeed an error. I changed the naming when I posted my code snippet here and was not doing the change consistently.

The scale_function is correctly defined as:

```python
    def scale_function(step):
           if step < 1000:
               scale = scale_factor[step]
           else:
              scale = 1
           return scale
```

**Comment by rob-hen** [comment]
Hi @awaelchli 

using your MWE, I was able to reproduce the bug when I print `self.optimizers().param_groups[0]['lr']:`

```python
import torch
from torch.optim.lr_scheduler import LambdaLR
from torch.utils.data import DataLoader, Dataset
import numpy as np
from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()

        print(
            f"lr: {self.trainer.lr_scheduler_configs[0].scheduler.get_last_lr()[0]:.4f}",
            f"lr_opt: {self.trainer.optimizers[0].param_groups[0]['lr']:.4f}",
            f"optimizers() lr:  {self.optimizers().param_groups[0]['lr']:.4f}"

        )
        return loss

    def on_validation_epoch_start(self):
        print("VALIDATION")

    def validation_step(self, batch, batch_idx):

        loss = self(batch).sum()
        return loss

    def configure_optimizers(self):
        print("configure call")

        scale_factor = np.linspace(1e-2, 1, num=16)
        optimizer = torch.optim.AdamW(self.layer.parameters(), lr=0.1)

        def scale_function(step):

            if step < 16:
                scale = scale_factor[step]
            else:
                scale = 1
            return scale
        scheduler = LambdaLR(optimizer, lr_lambda=scale_function)
        return [optimizer], [{"scheduler": scheduler, "interval": "step", "frequency": 1}]


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=8)

    model = BoringModel()
    trainer = Trainer(max_epochs=1, enable_model_summary=False,
                      enable_progress_bar=False)
    trainer.fit(model, train_dataloaders=train_data)

    print("Resume")
    model = BoringModel()
    trainer2 = Trainer(max_epochs=2, enable_model_summary=False,
                       enable_progress_bar=False)
    trainer2.fit(model, train_dataloaders=train_data,
                 ckpt_path=trainer.checkpoint_callback.best_model_path)


if __name__ == "__main__":
    run()
``` 

Output:
```
lr: 0.0010 lr_opt: 0.0010 optimizers() lr:  0.0010
lr: 0.0076 lr_opt: 0.0076 optimizers() lr:  0.0076
lr: 0.0142 lr_opt: 0.0142 optimizers() lr:  0.0142
lr: 0.0208 lr_opt: 0.0208 optimizers() lr:  0.0208
lr: 0.0274 lr_opt: 0.0274 optimizers() lr:  0.0274
lr: 0.0340 lr_opt: 0.0340 optimizers() lr:  0.0340
lr: 0.0406 lr_opt: 0.0406 optimizers() lr:  0.0406
lr: 0.0472 lr_opt: 0.0472 optimizers() lr:  0.0472

Resume

lr: 0.0538 lr_opt: 0.0538 optimizers() lr:  0.0010
lr: 0.0604 lr_opt: 0.0604 optimizers() lr:  0.0010
lr: 0.0670 lr_opt: 0.0670 optimizers() lr:  0.0010
lr: 0.0736 lr_opt: 0.0736 optimizers() lr:  0.0010
lr: 0.0802 lr_opt: 0.0802 optimizers() lr:  0.0010
lr: 0.0868 lr_opt: 0.0868 optimizers() lr:  0.0010
lr: 0.0934 lr_opt: 0.0934 optimizers() lr:  0.0010
lr: 0.1000 lr_opt: 0.1000 optimizers() lr:  0.0010

```


Can you explain what is the difference between `self.trainer.lr_scheduler_configs[0].scheduler.get_last_lr()[0]`,  `self.trainer.optimizers[0].param_groups[0]['lr']` and  `self.optimizers().param_groups[0]['lr']` and why the last one differs from the first two, when we resume from a checkpoint?

**Comment by awaelchli** [comment]
Thanks a lot @rob-hen. 
This actually confirms what was recently fixed in #18280. The fix was released in 2.0.7. If you install the latest lightning version, you will see it working. In fact, the learning rate was always loaded correctly, the bug is actually just a "display bug" with the optimizer wrapper returned by `self.optimizers()`. Thanks for the repro, and please let me know if the bugfix patch worked.
